addons/mail/static/src/models/message_action.js

```diff
@@ -0,0 +1,336 @@
+                    break;
+                default:
+                    break;
+            }
+            return classAttributes.join(' ');
+        },
+        /**
+         * @private
+         * @returns {FieldCommand}
+         */
+        _computeMessageActionListOwner() {
+            if (this.messageActionListOwnerAsDelete) {
+                return replace(this.messageActionListOwnerAsDelete);
+            } else if (this.messageActionListOwnerAsEdit) {
+                return replace(this.messageActionListOwnerAsEdit);
+            } else if (this.messageActionListOwnerAsMarkAsRead) {
+                return replace(this.messageActionListOwnerAsMarkAsRead);
+            } else if (this.messageActionListOwnerAsReaction) {
+                return replace(this.messageActionListOwnerAsReaction);
+            } else if (this.messageActionListOwnerAsReplyTo) {
+                return replace(this.messageActionListOwnerAsReplyTo);
+            } else if (this.messageActionListOwnerAsToggleCompact) {
+                return replace(this.messageActionListOwnerAsToggleCompact);
+            } else if (this.messageActionListOwnerAsToggleStar) {
+                return replace(this.messageActionListOwnerAsToggleStar);
+            }
+            return clear();
+        },
+        /**
+         * @private
+         * @returns {FieldCommand}
+         */
+        _computeMessageActionView() {
+            if (
+                this.messageActionListOwner &&
+                !this.messageActionListOwner.actionToggleCompact ||
+                this.messageActionListOwner.actionToggleCompact &&
+                !this.messageActionListOwner.isCompact
+            ) {
+                return insertAndReplace();
+            } else if (
+                this.messageActionListOwner &&
+                this.messageActionListOwner.actionToggleCompact &&
+                this.messageActionListOwner.isCompact
+            ) {
+                if (this.messageActionListOwner == this.messageActionListOwnerAsReaction ||
+                    this.messageActionListOwner == this.messageActionListOwnerAsToggleCompact) {
+                        return insertAndReplace();
+                    }
+                    return clear();
+            }
+            return clear();
+        },
+        /**
+         * @private
+         * @returns {string}
+         */
+        _computePositionAtMessageActionList() {
+            const {
+                length: l,
+                [0]: smallestequenceValue,
+                [l - 1]: largestSequenceValue,
+            } = Array(...this.messageActionListOwner.messageActions.map(action => action.sequence));
+            if (largestSequenceValue!== null && this.sequence >= largestSequenceValue) {
+                return 'right';
+            }
+            if (smallestequenceValue!== null && this.sequence <= smallestequenceValue) {
+                return 'left';
+            }
+            return 'mid';
+        },
```

Comment:
it's not good to have a compute where each action depend on each other action in a loop like that

Reply:
Also if I understand the goal correctly the proper modelling here should be depending on views (actions actually displayed), not all existing actions.

This should be moved to action list, where it should simply be computing first and last action from views.

`positionAtMessageActionList` could be removed in favor of checking whether the current action is the first or last one.

Reply:
Agree, just define computed fields `MessageActionList/firstAction` and `MessageActionList/lastAction`, each being in computational complexity `O(1)`.

Note that current implementation of `_computePositionAtMessageActionList` is computational complexity `O(n)` because there's a loop (i.e. `.map()` over all the message actions in the list).